File: VB/WebSite/App_Code/ColumnInfo.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.UI.WebControls
Imports System.Xml.Serialization
Imports DevExpress.Data
Imports DevExpress.Web.ASPxGridView

<XmlInclude(GetType(DataColumnInfo)), XmlInclude(GetType(CommandColumnInfo)), XmlInclude(GetType(BandColumnInfo))> _
Public MustInherit Class ColumnInfoBase
	Public Sub New()
	End Sub
	Public Sub New(ByVal column As GridViewColumn)
		Caption = column.Caption
		Width = column.Width
		Name = column.Name
		ExportWidth = column.ExportWidth
		VisibleIndex = column.VisibleIndex
		Visible = column.Visible
	End Sub

	Private privateVisibleIndex As Integer
	Public Property VisibleIndex() As Integer
		Get
			Return privateVisibleIndex
		End Get
		Set(ByVal value As Integer)
			privateVisibleIndex = value
		End Set
	End Property

	Private privateWidth As Unit
	Public Property Width() As Unit
		Get
			Return privateWidth
		End Get
		Set(ByVal value As Unit)
			privateWidth = value
		End Set
	End Property
	Private privateVisible As Boolean
	Public Property Visible() As Boolean
		Get
			Return privateVisible
		End Get
		Set(ByVal value As Boolean)
			privateVisible = value
		End Set
	End Property
	Private privateName As String
	Public Property Name() As String
		Get
			Return privateName
		End Get
		Set(ByVal value As String)
			privateName = value
		End Set
	End Property
	Private privateCaption As String
	Public Property Caption() As String
		Get
			Return privateCaption
		End Get
		Set(ByVal value As String)
			privateCaption = value
		End Set
	End Property
	Private privateExportWidth As Integer
	Public Property ExportWidth() As Integer
		Get
			Return privateExportWidth
		End Get
		Set(ByVal value As Integer)
			privateExportWidth = value
		End Set
	End Property



	Public Overridable Sub RestoreGridViewColumn(ByVal column As GridViewColumn)
		column.Caption = Caption
		column.Width = Width
		column.Name = Name
		column.ExportWidth = ExportWidth
		column.VisibleIndex = VisibleIndex
		column.Visible = Visible
	End Sub
End Class

Public Class DataColumnInfo
	Inherits ColumnInfoBase
	Public Sub New()
	End Sub
	Public Sub New(ByVal column As GridViewDataColumn)
		MyBase.New(column)
		RealTypeName = column.GetType().ToString()
		FieldName = column.FieldName
		GroupIndex = column.GroupIndex
		SortIndex = column.SortIndex
		SortOrder = column.SortOrder
	End Sub

	Private privateRealTypeName As String
	Public Property RealTypeName() As String
		Get
			Return privateRealTypeName
		End Get
		Set(ByVal value As String)
			privateRealTypeName = value
		End Set
	End Property
	Private privateFieldName As String
	Public Property FieldName() As String
		Get
			Return privateFieldName
		End Get
		Set(ByVal value As String)
			privateFieldName = value
		End Set
	End Property
	Private privateGroupIndex As Integer
	Public Property GroupIndex() As Integer
		Get
			Return privateGroupIndex
		End Get
		Set(ByVal value As Integer)
			privateGroupIndex = value
		End Set
	End Property
	Private privateSortIndex As Integer
	Public Property SortIndex() As Integer
		Get
			Return privateSortIndex
		End Get
		Set(ByVal value As Integer)
			privateSortIndex = value
		End Set
	End Property
	Private privateSortOrder As ColumnSortOrder
	Public Property SortOrder() As ColumnSortOrder
		Get
			Return privateSortOrder
		End Get
		Set(ByVal value As ColumnSortOrder)
			privateSortOrder = value
		End Set
	End Property

	Public Overrides Sub RestoreGridViewColumn(ByVal column As GridViewColumn)
		If column Is Nothing Then
			Return
		End If
		If GetType(GridViewDataColumn).Assembly.GetType(RealTypeName) IsNot column.GetType() Then
			Return
		End If

		MyBase.RestoreGridViewColumn(column)
		Dim c As GridViewDataColumn = CType(column, GridViewDataColumn)
		c.FieldName = FieldName
		c.GroupIndex = GroupIndex
		c.SortIndex = SortIndex
		c.SortOrder = SortOrder
	End Sub
End Class

Public NotInheritable Class CommandColumnInfo
	Inherits ColumnInfoBase
	Public Sub New()
	End Sub
	Public Sub New(ByVal column As GridViewCommandColumn)
		MyBase.New(column)
		ShowEditButton = column.EditButton.Visible
		ShowNewButton = column.NewButton.Visible
		ShowEditButton = column.EditButton.Visible
		ShowSelectCheckbox = column.ShowSelectCheckbox
	End Sub

	Private privateShowEditButton As Boolean
	Public Property ShowEditButton() As Boolean
		Get
			Return privateShowEditButton
		End Get
		Set(ByVal value As Boolean)
			privateShowEditButton = value
		End Set
	End Property
	Private privateShowNewButton As Boolean
	Public Property ShowNewButton() As Boolean
		Get
			Return privateShowNewButton
		End Get
		Set(ByVal value As Boolean)
			privateShowNewButton = value
		End Set
	End Property
	Private privateShowDeleteButton As Boolean
	Public Property ShowDeleteButton() As Boolean
		Get
			Return privateShowDeleteButton
		End Get
		Set(ByVal value As Boolean)
			privateShowDeleteButton = value
		End Set
	End Property
	Private privateShowSelectCheckbox As Boolean
	Public Property ShowSelectCheckbox() As Boolean
		Get
			Return privateShowSelectCheckbox
		End Get
		Set(ByVal value As Boolean)
			privateShowSelectCheckbox = value
		End Set
	End Property

	Public Overrides Sub RestoreGridViewColumn(ByVal column As GridViewColumn)
		If column Is Nothing Then
			Return
		End If
		MyBase.RestoreGridViewColumn(column)
		Dim c As GridViewCommandColumn = CType(column, GridViewCommandColumn)
		c.EditButton.Visible = ShowEditButton
		c.NewButton.Visible = ShowNewButton
		c.EditButton.Visible = ShowEditButton
		c.ShowSelectCheckbox = ShowSelectCheckbox
	End Sub
End Class

Public NotInheritable Class BandColumnInfo
	Inherits ColumnInfoBase
	Public Sub New()
	End Sub
	Public Sub New(ByVal column As GridViewBandColumn)
		MyBase.New(column)
	End Sub

	Public Overrides Sub RestoreGridViewColumn(ByVal column As GridViewColumn)
		If column Is Nothing Then
			Return
		End If
		MyBase.RestoreGridViewColumn(column)
		Dim c As GridViewBandColumn = CType(column, GridViewBandColumn)
		'To-Do: Save child column information		
	End Sub
End Class

File: CS/WebSite/App_Code/ColumnInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.UI.WebControls;
using System.Xml.Serialization;
using DevExpress.Data;
using DevExpress.Web;

[XmlInclude(typeof(DataColumnInfo))]
[XmlInclude(typeof(CommandColumnInfo))]
[XmlInclude(typeof(BandColumnInfo))]
public abstract class ColumnInfoBase {
	public ColumnInfoBase() {
	}
	public ColumnInfoBase(GridViewColumn column) {
		Caption = column.Caption;
		Width = column.Width.ToString();
		Name = column.Name;
		ExportWidth = column.ExportWidth;
		VisibleIndex = column.VisibleIndex;
		Visible = column.Visible;
	}

	public int VisibleIndex { get; set; }

	public string Width { get; set; }
	public bool Visible { get; set; }
	public string Name { get; set; }
	public string Caption { get; set; }
	public int ExportWidth { get; set; }

	

	public virtual void RestoreGridViewColumn(GridViewColumn column) {
		column.Caption = Caption;
		column.Width = Unit.Parse(Width);
		column.Name = Name;
		column.ExportWidth = ExportWidth;
		column.VisibleIndex = VisibleIndex;
		column.Visible = Visible;
	}
}

public class DataColumnInfo : ColumnInfoBase {
	public DataColumnInfo() {
	}
	public DataColumnInfo(GridViewDataColumn column)
		: base(column) {
		RealTypeName = column.GetType().ToString();
		FieldName = column.FieldName;
		GroupIndex = column.GroupIndex;
		SortIndex = column.SortIndex;
		SortOrder = column.SortOrder;
	}

	public string RealTypeName { get; set; }
	public string FieldName { get; set; }
	public int GroupIndex { get; set; }
	public int SortIndex { get; set; }
	public ColumnSortOrder SortOrder { get; set; }

	public override void RestoreGridViewColumn(GridViewColumn column) {
		if (column == null)
			return;
		if (typeof(GridViewDataColumn).Assembly.GetType(RealTypeName) != column.GetType())
			return;

		base.RestoreGridViewColumn(column);
		GridViewDataColumn c = (GridViewDataColumn)column;		
		c.FieldName = FieldName;
		c.GroupIndex = GroupIndex;
		c.SortIndex = SortIndex;
		c.SortOrder = SortOrder;		
	}
}

public sealed class CommandColumnInfo : ColumnInfoBase {
	public CommandColumnInfo() {
	}
	public CommandColumnInfo(GridViewCommandColumn column)
		: base(column) {
		ShowEditButton = column.ShowEditButton;
		ShowNewButton = column.ShowSelectButton;
		ShowDeleteButton = column.ShowDeleteButton;
		ShowSelectCheckbox = column.ShowSelectCheckbox;
	}
		
	public bool ShowEditButton { get; set; }
	public bool ShowNewButton { get; set; }
	public bool ShowDeleteButton { get; set; }
	public bool ShowSelectCheckbox { get; set; }

	public override void RestoreGridViewColumn(GridViewColumn column) {
		if (column == null)
			return;
		base.RestoreGridViewColumn(column);
		GridViewCommandColumn c = (GridViewCommandColumn)column;		
		c.ShowEditButton = ShowEditButton;
		c.ShowNewButton = ShowNewButton;
		c.ShowDeleteButton = ShowDeleteButton;
		c.ShowSelectCheckbox = ShowSelectCheckbox;		
	}
}

public sealed class BandColumnInfo : ColumnInfoBase {
	public BandColumnInfo() {
	}
	public BandColumnInfo(GridViewBandColumn column)
		: base(column) {
	}

	public override void RestoreGridViewColumn(GridViewColumn column) {
		if (column == null)
			return;
		base.RestoreGridViewColumn(column);
		GridViewBandColumn c = (GridViewBandColumn)column;
		//To-Do: Save child column information		
	}
}